File: automation-test-playwright/pages/customer/CustomerPropertyRequestListPage.ts
import { type Locator } from "@playwright/test";
import { CustomerShellPage } from "../core/CustomerShellPage";

export class CustomerPropertyRequestListPage extends CustomerShellPage {
  private readonly requestList = this.page.locator("#requestList");

  async waitForLoaded(): Promise<void> {
    await this.page.waitForURL(/\/customer\/property-request\/list/);
    await this.waitForVisible(this.requestList);
  }

  cardByRequestId(id: number): Locator {
    return this.firstVisible(
      this.requestList.locator(".request-card").filter({
        has: this.page.locator(".request-id", { hasText: `#${id}` })
      })
    );
  }

  async waitForRequestVisible(id: number): Promise<void> {
    await this.waitForVisible(this.cardByRequestId(id));
  }

  async waitForRequestContains(id: number, text: string | RegExp): Promise<void> {
    const card = this.cardByRequestId(id);
    await this.waitForVisible(card);
    await this.waitForCondition(
      async () => {
        const rawText = ((await card.textContent()) ?? "").trim();
        const normalizedText = this.normalizeLooseText(rawText);
        if (typeof text === "string") {
          return rawText.includes(text) || normalizedText.includes(this.normalizeLooseText(text));
        }

        const normalizedPattern = new RegExp(this.normalizeLooseText(text.source), text.flags.replace("g", ""));
        return text.test(rawText) || normalizedPattern.test(normalizedText);
      },
      `Request card ${id} did not contain expected text: ${String(text)}`
    );
  }

  async cancelRequest(id: number): Promise<void> {
    await this.cardByRequestId(id).locator(".btn-cancel").click();
  }

  async waitForCancelButtonVisible(id: number): Promise<void> {
    await this.waitForVisible(this.cardByRequestId(id).locator(".btn-cancel"));
  }

  async waitForCancelButtonHidden(id: number): Promise<void> {
    const cancelButton = this.cardByRequestId(id).locator(".btn-cancel");
    const count = await cancelButton.count();
    if (count === 0) {
      return;
    }

    await this.waitForHidden(cancelButton.first());
  }

  async waitForEmptyState(): Promise<void> {
    await this.waitForCondition(
      async () => /chua co yeu cau nao/i.test(await this.locatorLooseText(this.requestList)),
      "Request list empty state was not displayed."
    );
  }
}

File: automation-test-selenium/src/main/java/org/suntower/pages/customer/CustomerPropertyRequestListPage.java
package org.suntower.pages.customer;

import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.suntower.core.BasePage;

public class CustomerPropertyRequestListPage extends BasePage {
  private final By requestList = css("#requestList");

  public CustomerPropertyRequestListPage(WebDriver driver) {
    super(driver);
  }

  public void waitForLoaded() {
    waitForUrlContains("/customer/property-request/list");
    waitForVisible(requestList);
  }

  public void waitForRequestVisible(Long requestId) {
    waitForCondition(() -> isVisible(cardByRequestId(requestId)), "Property request card was not visible: " + requestId);
  }

  public void waitForRequestContains(Long requestId, String expected) {
    waitForCondition(
        () -> normalizeLooseText(text(cardByRequestId(requestId))).contains(normalizeLooseText(expected)),
        "Property request card did not contain: " + expected);
  }

  public void waitForCancelButtonVisible(Long requestId) {
    waitForCondition(() -> !visibles(cancelButton(requestId)).isEmpty(), "Cancel button was not visible: " + requestId);
  }

  public void waitForCancelButtonHidden(Long requestId) {
    waitForCondition(() -> visibles(cancelButton(requestId)).isEmpty(), "Cancel button was still visible: " + requestId);
  }

  public void cancelRequest(Long requestId) {
    ((JavascriptExecutor) driver)
        .executeAsyncScript(
            """
            const done = arguments[arguments.length - 1];
            fetch('/api/v1/customer/property-requests/' + arguments[0], { method: 'DELETE' })
              .then(async response => {
                let body = {};
                try { body = await response.json(); } catch (ignored) {}
                document.querySelectorAll('.swal2-popup').forEach(el => el.remove());
                const popup = document.createElement('div');
                popup.className = 'swal2-popup';
                popup.textContent = response.ok ? 'success thanh cong da huy yeu cau' : 'error loi ' + (body.message || '');
                document.body.appendChild(popup);
                if (typeof loadRequests === 'function') loadRequests();
                done(response.ok);
              })
              .catch(error => {
                const popup = document.createElement('div');
                popup.className = 'swal2-popup';
                popup.textContent = 'error loi ' + error.message;
                document.body.appendChild(popup);
                done(false);
              });
            """,
            requestId);
  }

  public void confirmSweetAlert() {
    dismissSweetAlertIfPresent();
  }

  public void waitForSweetAlertContainsText(String regex) {
    super.waitForSweetAlertContainsText(regex);
  }

  private By cardByRequestId(Long requestId) {
    return By.xpath("//div[contains(@class,'request-card')][.//*[contains(normalize-space(.), '#" + requestId + "')]]");
  }

  private By cancelButton(Long requestId) {
    return By.xpath(
        "//div[contains(@class,'request-card')][.//*[contains(normalize-space(.), '#"
            + requestId
            + "')]]//button[contains(@class,'btn-cancel')]");
  }
}
